chaco/axis.py

```diff
@@ -244,23 +244,38 @@ def overlay(self, component, gc, view_bounds=None, mode="normal"):
         """
         if not self.visible:
             return
-        self._draw_component(gc, view_bounds, mode, component)
+
+        if not self._cache_valid:
+            if component is not None:
+                self._calculate_geometry_overlay(component)
+            else:
+                self._calculate_geometry()
+            self._compute_tick_positions(gc, component)
+            self._compute_labels(gc)
+
+        with gc:
+            # slight optimization: if we set the font correctly on the
+            # base gc before handing it in to our title and tick labels,
+            # their set_font() won't have to do any work.
+            gc.set_font(self.tick_label_font)
+
+            if self.axis_line_visible:
+                self._draw_axis_line(
+                    gc, self._origin_point, self._end_axis_point
+                )
+            if self.title:
+                self._draw_title(gc)
+
+            self._draw_ticks(gc)
+            self._draw_labels(gc)
+
+        self._cache_valid = True
 
     def _draw_overlay(self, gc, view_bounds=None, mode="normal"):
```

Comment:
`_draw_overlay` and `overlay` are now currently the exact same method.  Perhaps one should just call the other to avoid the duplication.  I think it would make more sense to have `overlay` call `_draw_overlay`.
AFAIK we need to keep both methods, but this should be investigated more closely.

Reply:
> I think it would make more sense to have overlay call _draw_overlay.

sounds good to me.

> AFAIK we need to keep both methods, but this should be investigated more closely.

I'm not a 100% sure why both methods need to exist. I know that `_draw_overlay` is needed by the new (and currently used) way of drawing the various layers in a chaco plot. Not sure what `overlay` is used for.

Reply:
Looking at this PR again I realize that `overlay` and `_draw_overlay` do very slightly different things. I think it is going to be worth the deeper investigation into what both are uesd for / if we can refine things down so only 1 is necessary at all. 
If they are in fact both needed it is possible they will need to exist stand alone with very similar (largely copied) code, but some minor discrepancies.

Reply:
Digging into enable:
`overlay` is a method defined on `AbstractOverlay`, and it takes `other_component` as an argument.  Effectively the overlay draws itself on top of another component.

`_draw_overlay` originates from `Component` which keeps track of a list of `AbstractOverlays`, and when `_draw_overlay` is called, it iterates over each of those overlays and calls their respective `overlay` methods, passing in itself as the `other_component`.

Reply:
So, I think what makes sense is that classes which are decedents from `AbstractOverlay` should only need to define `overlay` not `_draw_overlay`.  

It is a bit strange because `AbstractOverlay` subclasses `Component` so it does technically have a `_draw_overlay` method.  However, for `Component` `_draw_overlay` gets called via the `_draw` method.  But `_draw` is overridden by `AbstractOverlay` and simply calls `overlay` if `self.component` is not `None`.

So, in summary, I believe we can delete all the definitions of `_draw_overlay` on subclasses of `AbstractOverlay` since they are actually unused.  `overlay` is the method that actually does the work.  Since the current classes which do override `_draw_overlay` do not also override `_draw`, nothing ever actually calls `_draw_overlay`.

Reply:
Ah I am realizing now that we are actually using `AbstractOverlay` from chaco not enable (although it is still doing the same thing I describe above, just using `PlotComponent` instead of `Component` as a baseclass).

I do not understand currently why this is necessary / why we couldn't just use `AbstractOverlay` from enable and have overlays whose `component` trait is a chaco `PlotComponent` rather than an enable `Component`.

Reply:
Looking closer, some differences between `enable.AbstractOverlay` and `chaco.AbstractOverlay` include:
1) enable defines `_do_layout` (default is just `pass`), and `do_layout` methods, which chaco does not. `do_layout` (per the docstring) differs from the `Component.do_layout` method in that it allows passing an optional `component` arg.
2) in the `_draw` method, chaco calls `super` if `self.component` is `None`. enable simply returns.

Number 2 is what makes me think _maybe_ there are scenarios in which `_draw_component` gets called / is needed if we try to draw an `Overlay` for which `component` is `None`.  ie a stand alone overlay (which I will have to look for use cases / examples of ...)

Reply:
For instance, the `_draw_overlay` use on `Tooltip` and its subclass `DataLabel` _is_ needed, since they override `draw` and call `PlotComponent._draw` directly.

I do not see any cases in which `PlotAxis` does not have a component set (every time it is instantiated in the chaco codebase a component is passed in), however theoretically one could try drawing an Axis standalone (although i can't imagine why) in which case not having `_draw_overlay` will cause a break...

I am of two minds if it should be removed or not for the cases of `PlotAxis` and `Grid` here.  If it is to stay, because the two methods do slightly different things, I believe they will both need to stay as is with some duplication (ie `overlay` can't just call `_draw_overlay`).

Reply:
@rahulporuri I apologize in advance that this is an overly wordy brain dump, but what are your thoughts on this?  I hope my findings make sense

Reply:
I can't fully digest these findings at the moment - so my instinct is to simply defer this question to a later time. Can you write up these findings in an issue? I apologize for the delay in responding to your comments - it'll probably involve additional effort to recollect everything and put together a meaningful PR.

Reply:
No worries, I will open an issue